match.py

```diff
@@ -26,92 +26,28 @@ def parse_args():
-            household_project = Project(
-                project_name,
-                household_schema,
-                c.systems,
-                c.entity_service_url,
-                c.blocked,
-            )
-            household_project.start_project()
-            for system in c.systems:
-                household_project.upload_clks(
-                    system, c.get_household_clks_raw(system, project_name)
-                )
-            household_project.start_run(c.matching_threshold)
-            running = True
-            while running:
-                status = household_project.get_run_status()
-                print(status)
-                if status.get("state") == "completed":
-                    running = False
-                    break
-                time.sleep(SLEEP_TIME)
-            result_json = household_project.get_results()
-            results = Results(c.systems, project_name, result_json)
-            results.insert_results(database.household_match_groups)
-    else:
-        log.debug("Processing individuals")
-        if c.blocked:
-            log.debug("Blocked, extracting CLKs and blocks")
-            for system in c.systems:
-                c.extract_clks(system)
-                c.extract_blocks(system)
-
-        iter_num = 0
-        for project_name, schema in c.load_schema().items():
-            iter_num = iter_num + 1
-            project = Project(
-                project_name, schema, c.systems, c.entity_service_url, c.blocked
-            )
-            project.start_project()
-            for system in c.systems:
-                if c.blocked:
-                    project.upload_clks_blocked(
-                        system,
-                        c.get_clk(system, project_name),
-                        c.get_block(system, project_name),
-                    )
-                else:
-                    project.upload_clks(system, c.get_clks_raw(system, project_name))
-            project.start_run(c.matching_threshold)
-            running = True
-            print("\n--- RUNNING ---\n")
-            while running:
-                status = project.get_run_status()
-                print(status)
-                if status.get("state") == "completed":
-                    running = False
-                    break
-                time.sleep(SLEEP_TIME)
-            print("\n--- Getting results ---\n")
-            result_json = project.get_results()
-            results = Results(c.systems, project_name, result_json)
-            print(
-                "Matching groups for system "
-                + str(iter_num)
-                + " of "
-                + str(len(c.load_schema().items()))
-            )
-            results.insert_results(database.match_groups)
+    return parser.parse_args()
+
+def do_match(config):
+    database = MongoClient(config.mongo_uri).linkage_agent
+    for file_name in Path(config.project_results_dir).glob('*.json'):
```

Comment:
Quick thought - this will still run even if not all projects completed successfully in the previous step. I expect the user will catch it if something went wrong, but it wouldn't be ideal if a project failed and then the process continued on without it.  Can this instead loop over `config.projects` and add .json to get the filename, where it could fail if an expected project was missing?

Reply:
Good call and I like the suggested approach.

Reply:
Implemented in https://github.com/mitre/linkage-agent-tools/pull/21/commits/74de8fc409f7ac5c1b9a214336c6ecabf325d0e9

When results are missing `./match.py` will throw an exception indicating the missing projects:

```shell
  File "linkage-agent-tools/./match.py", line 71, in <module>
    do_match(config)
  File "linkage-agent-tools/./match.py", line 50, in do_match
    has_results_available(config)
  File "linkage-agent-tools/./match.py", line 46, in has_results_available
    raise MissingResults(expected_results, available_results)
__main__.MissingResults: Missing results for projects: name-sex-dob-zip, name-sex-dob-addr
```

If all the files are present then matching will proceed normally:

```shell
Matching groups for system name-sex-dob-phone
Inserting 1477 records into the local database: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1477/1477 [00:03<00:00, 391.70it/s]
Matching groups for system name-sex-dob-zip
Inserting 1442 records into the local database: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1442/1442 [00:03<00:00, 383.96it/s]
Matching groups for system name-sex-dob-addr
Inserting 1512 records into the local database: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1512/1512 [00:03<00:00, 406.36it/s]
```

I'll work on adding some unit tests for this